File: wee_mpi/examples/fortran/ring.f90
program ring
    use MPI
    use wee_mpi 
    use iso_c_binding, only: C_CHAR, C_NULL_CHAR

    implicit none

    integer, parameter :: comm=MPI_COMM_WORLD
    integer :: rank, len, size, ierr, token, istat(MPI_STATUS_SIZE)
    character*(MPI_MAX_PROCESSOR_NAME) name
    call MPI_Init(ierr)
    call MPI_Comm_size(comm, size, ierr)
    call MPI_Comm_rank(comm, rank, ierr)
    call MPI_GET_PROCESSOR_NAME(name, len, ierr)
    
    ierr = f_wee_anim("arrow-up")  

    print *, "Rank, Name", rank, name
    if (rank .eq. 0) token = 1 
    if (rank .ne. 0 ) then 
        call MPI_Recv(token, 1, MPI_INTEGER, rank - 1, 0, comm, istat, ierr)
        print *, "Process ", rank, " received ", token, " from process ", rank-1 
    endif 
    
    call MPI_SSend(token, 1, MPI_INTEGER, modulo(rank + 1, size), 0, comm, ierr)

    if (rank .eq. 0) then
        call mpi_recv(token, 1, mpi_integer, size - 1, 0, comm, istat, ierr)
        print *, "process ", rank, " received ", token, " from process ", size-1 
    endif

    call MPI_Finalize(ierr)
end 

File: wee_mpi/examples/ring.c
#include <stdio.h>
#include <stdlib.h>
#include <wee_mpi.h>


int main(int argc, char** argv) {
    // initialize the mpi environment
    MPI_Init(&argc, &argv);
 
    // get the number of processes
    int world_size;
    MPI_Comm_size(MPI_COMM_WORLD, &world_size);

    // get the rank of the process
    int world_rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
    
    // get the name of the processor
    char processor_name[MPI_MAX_PROCESSOR_NAME]; int name_len;
    MPI_Get_processor_name(processor_name, &name_len);

    printf("Rank: %d, Proc Name:%s\n", world_rank, processor_name); 
   	int token;
	if (world_rank != 0) {
		MPI_Recv(&token, 1, MPI_INT, world_rank - 1, 0,
				 MPI_COMM_WORLD, MPI_STATUS_IGNORE);
		printf("Process %d received token %d from process %d\n",
			   world_rank, token, world_rank - 1);
	} else {
		// Set the token's value if you are process 0
		token = -1;
	}
	MPI_Ssend(&token, 1, MPI_INT, (world_rank + 1) % world_size,
			 0, MPI_COMM_WORLD);

	// Now process 0 can receive from the last process.
	if (world_rank == 0) {
		MPI_Recv(&token, 1, MPI_INT, world_size - 1, 0,
				 MPI_COMM_WORLD, MPI_STATUS_IGNORE);
		printf("Process %d received token %d from process %d\n",
			   world_rank, token, world_size - 1);
	} 

    // Finalize the MPI environment.
    MPI_Finalize();
}
